src/ClientData/include/ClientData/ScopeInfo.h

```diff
@@ -0,0 +1,30 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_CLIENT_DATA_SCOPE_INFO_H_
+#define ORBIT_CLIENT_DATA_SCOPE_INFO_H_
+
+#include <string>
+
+namespace orbit_client_data {
+
+enum ScopeType : uint8_t { kOther = 0, kHookedFunction = 1, kApiScope = 2, kApiScopeAsync = 3 };
+
+struct ScopeInfo {
```

Comment:
I would personally prefer a class with constructor and getters, making the type immutable, which is generally a good thing.

Reply:
Do you suggest to add const modifiers to its fields, thus making it not-movable? Why isn't adding const modifier to its instances enough? E. g. `GetScopeInfo` returns const-ref. I will also change the type of `NameEqualityScopeIdProvider::scope_id_to_info_` to be `absl::flat_hash_map<uint64_t, *const* ScopeInfo>`.

Reply:
> Do you suggest to add const modifiers to its fields, thus making it not-movable

No, I suggest making the fields private and adding getters.

> Why isn't adding const modifier to its instances enough?

Because that's no longer a property of the class.

Reply:
I see your point. And I would agree with you, if we didn't have const semantics. If we make the fields private, we pay with boilerplate, but I don't see what we gain. Implicit immutability?

Reply:
- There is a big difference between making the type immutable and saying that `const` should be used everywhere where the type is used. Public fields here are equivalent to having getters and setters. If you don't need the setters (which you don't), only define the getters.
- There is no boilerplate issue, you just use the getters instead of the fields directly.

Reply:
> There is no boilerplate issue

But I still need to define the getters. With struct I've got these for free. 

> and saying that const should be used everywhere where the type is used

I don't say that. Maybe somewhere it will be convenient to  write smth like that. 

```
ScopeInfo scope_info; // meaningless state
scope_info.name = foo(); 
if (cond) {
   scope_info.type = boo()
}
scope_info.type = bar(); 
```

Moreover, the need to write `const` if we need a const instance makes intention explicit. The reader wouldn't have to go back to the definition and be like "oh, it has no setters, it's immutable".

Reply:
> But I still need to define the getters. With struct I've got these for free.

Come on, this is really a non-argument. It's like three lines of code on class definition. And it improves the semantics.

> ScopeInfo scope_info; // meaningless state

That's my point (I also said it at the beginning with having a proper constructor). I don't like the meaningless state. No mutability means no meaningless or intermediate state.

Reply:
> And it improves the semantics.

Debatable. Immutability will not as clear to the reader. 

> I don't like the meaningless state

Then you've got the freedom of using `const ScopeInfo`. 

I think we need the third opinion. @beckerhe @florian-kuebler @vickyliu-go4it ?

Reply:
I am with @dpallotti.

Reply:
Also, do you suggest to remove the default constructor? Wouldn't it lead to problems with storing it in containers?

Reply:
Yes, I don't think there should be a default constructor. Storing it in a container is mostly fine. *some* functions would be available though.

Reply:
Then so be it.